distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Text_Sub_Range.enso

```diff
@@ -0,0 +1,123 @@
+
+polyglot java import com.ibm.icu.text.BreakIterator
+polyglot java import org.enso.base.Text_Utils
+
+## Type defining a substring of a Text
+type Text_Sub_Range
+    ## Select the first `count` characters.
+       Select an empty string if `count` is less than or equal to 0.
+       Select the entire string if `count` is greater than the length of the input.
+    type First (count : Integer = 1)
+
+    ## Select the last `count` characters.
+       Select an empty string if `count` is less than or equal to 0.
+       Select the entire string if `count` is greater than the length of the input.
+    type Last (count : Integer = 1)
+
+    ## Select characters until the first instance of `delimiter`.
+       Select an empty string if `delimiter` is empty.
+       Select the entire string if the input does not contain `delimiter`.
+    type Before (delimiter : Text)
+
+    ## Select characters until the last instance of `delimiter`.
+       Select an empty string if `delimiter` is empty.
+       Select the entire string if the input does not contain `delimiter`.
+    type Before_Last (delimiter : Text)
+
+    ## Select characters after the first instance of `delimiter`.
+       Select an empty string if the input does not contain `delimiter`.
+    type After (delimiter : Text)
+
+    ## Select characters after the last instance of `delimiter`.
+       Select an empty string if the input does not contain `delimiter`.
+    type After_Last (delimiter : Text)
+
+    ## Select characters while the predicate returns `True`.
+    type While (predicate : (Text -> Boolean))
+
+    ## PRIVATE
+       Evaluates the Text_Sub_Range returning the underlying char array indices
+    to_char_range : Text -> Range
+    to_char_range text =
+
+        ## Utility function to find char indices for Text_Sub_Range.
+           Arguments:
+           - text: Text to search
+           - predicate: Function to test each character, receives:
+             - index: current index
+               start: index the char array to start of grapheme cluster
+               end: index the char array to start of next grapheme cluster
+             return True to exit loop
+           Returns: either a Pair of char indices for current grapheme cluster or
+             Pair -1 (char array length) if not found.
+        find_sub_range_end = text->predicate->
+            iterator = BreakIterator.getCharacterInstance
+            iterator.setText text
+
+            loop index start end =
+                if end == -1 then (Pair -1 start) else
+                    if predicate index start end then (Pair start end) else
+                        @Tail_Call loop (index + 1) end iterator.next
+
+            loop 0 0 iterator.next
+
+        case this of
+            First count ->
+                if count <= 0 then (Range 0 0) else
+                    indices = find_sub_range_end text (index->_->_-> index+1 == count)
+                    Range 0 indices.second
```

Comment:
Maybe instead of this iteration with a lot of callbacks we could do `breakIterator.next count` ? Sounds like it could be slightly faster and more concise.

Same could go for last - `breakIterator.last(); breakIterator.prev(count)`.

Reply:
adding ticket to review the BreakIterators https://www.pivotaltracker.com/story/show/181342391